community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/predicates/Equivalent.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_0.commands.predicates
+
+import org.neo4j.cypher.internal.compiler.v3_0.{CRS, GeographicPoint}
+import org.neo4j.graphdb.{Node, Path, Relationship}
+
+import scala.collection.GenTraversableOnce
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+// Class that calculates if two values are equivalent or not.
+// Does not handle NULL values - that must be handled outside!
+class Equivalent(protected val eagerizedValue: Any, val originalValue: Any) extends scala.Equals {
+  override def equals(in: Any): Boolean = {
+    val eagerOther = in match {
+      case s: Equivalent =>
+        if(originalValue.isInstanceOf[AnyRef] &&
+          s.originalValue.isInstanceOf[AnyRef] &&
+          (originalValue.asInstanceOf[AnyRef] eq s.originalValue.asInstanceOf[AnyRef])) return true
+        s.eagerizedValue
+      case x =>
+        if(originalValue.isInstanceOf[AnyRef] &&
+          x.isInstanceOf[AnyRef] &&
+          (originalValue.asInstanceOf[AnyRef] eq x.asInstanceOf[AnyRef])) return true
+        Equivalent.eager(x)
+    }
+
+    (eagerizedValue, eagerOther) match {
+      case (null, null) => true
+      case (n1: Double, n2: Float) => mixedFloatEquality(n2, n1)
+      case (n1: Float, n2: Double) => mixedFloatEquality(n1, n2)
+      case (a, b) => a == b
+      case _ => false
+    }
+  }
+
+  override def canEqual(that: Any): Boolean = true
+
+  private def mixedFloatEquality(a: Float, b: Double) =
+    a.doubleValue() == b.doubleValue() ||
+      (Math.rint(a.doubleValue()).toLong == b.longValue() &&
+        Math.rint(b.doubleValue()).toLong == a.longValue())
+
+  private var hash: Option[Int] = None
+
+  override def hashCode(): Int = hash.getOrElse {
+    val result = hashCode(eagerizedValue)
+    hash = Some(result)
+    result
+  }
+
+  private val EMPTY_LIST = 42
+
+  private def hashCode(o: Any): Int = o match {
+    case null => 0
+    case n: Number => n.longValue().hashCode()
```

Comment:
The Long zero has the hash code zero, so all numbers that go to a zero long will have the same hash code as `null` here. Given that zero might be the most commonly used number, wouldn't it make sense to give `null` another hash code, like `-1`, or `Int.MAX_VALUE`?